File: files/CH-034.R
library(tidyverse)
library(readxl)

input = read_excel("files/CH-034 Customer Return Cycle.xlsx", range = "B2:F26")
test  = read_excel("files/CH-034 Customer Return Cycle.xlsx", range = "J2:K6")

result = input %>%
  select(Date, `Customer ID`) %>%
  arrange(`Customer ID`, Date) %>%
  distinct() %>% 
  mutate(lag = lag(Date), .by = `Customer ID`) %>%
  mutate(diff = Date - lag) %>%
  summarise(`Avg Return Cycle` = mean(diff, na.rm = TRUE), .by = `Customer ID`) %>% 
  mutate(`Avg Return Cycle` = as.numeric(`Avg Return Cycle`)) %>%
  select(Customer = `Customer ID`, `Avg Return Cycle`)

identical(result, test)
# [1] TRUE

File: files/CH-034.py
import pandas as pd

input = pd.read_excel("CH-034 Customer Return Cycle.xlsx", sheet_name="Sheet1", usecols="B:F", skiprows=1)
test = pd.read_excel("CH-034 Customer Return Cycle.xlsx", sheet_name="Sheet1", usecols="J:K", skiprows=1, nrows = 4)

result = input[['Date', 'Customer ID']].sort_values(by=['Customer ID', 'Date']).drop_duplicates().reset_index(drop=True)
result['lag'] = result.groupby('Customer ID')['Date'].shift(1)
result['diff'] = (result['Date'] - result['lag']).dt.days
result = result.groupby('Customer ID')['diff'].mean().astype(int).reset_index()
result.columns = ['Customer', 'Avg Return Cycle']

print(result == test)   # True